### ISSUE TITLE
throw NPE when upload file from local to pod

### ISSUE THREAD

**Issue by wwulfric** [fabric8io/kubernetes-client#3477]
version: 5.7.3

```
2021-09-17 22:20:12,858 ERROR [pool-42-thread-1] utils.InputStreamPumper - Error while pumping stream.
java.lang.NullPointerException
        at io.fabric8.kubernetes.client.utils.InputStreamPumper$1.read(InputStreamPumper.java:53)
        at io.fabric8.kubernetes.client.utils.InputStreamPumper.transferTo(InputStreamPumper.java:76)
        at io.fabric8.kubernetes.client.utils.InputStreamPumper.lambda$pump$0(InputStreamPumper.java:89)
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1640)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2021-09-17 22:20:12,858 ERROR [pool-41-thread-1] utils.InputStreamPumper - Error while pumping stream.
java.lang.NullPointerException
        at io.fabric8.kubernetes.client.utils.InputStreamPumper$1.read(InputStreamPumper.java:53)
        at io.fabric8.kubernetes.client.utils.InputStreamPumper.transferTo(InputStreamPumper.java:76)
        at io.fabric8.kubernetes.client.utils.InputStreamPumper.lambda$pump$0(InputStreamPumper.java:89)
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1640)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2021-09-17 22:20:13,326 ERROR [pluginExecutor1] impl.K8sTunnel - [64535482163188841083311665968] [text={"metadata":{},"status":"Success"}]
io.fabric8.kubernetes.client.KubernetesClientException: [text={"metadata":{},"status":"Success"}]
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUploadWebSocketListener.checkError(PodUploadWebSocketListener.java:91)
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUploadWebSocketListener.waitUntilComplete(PodUploadWebSocketListener.java:111)
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUpload.uploadFile(PodUpload.java:87)
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUpload.upload(PodUpload.java:64)
        at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.lambda$upload$0(PodOperationsImpl.java:389)
        at io.fabric8.kubernetes.client.utils.OptionalDependencyWrapper.wrapRunWithOptionalDependency(OptionalDependencyWrapper.java:39)
        at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.upload(PodOperationsImpl.java:387)
        at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.upload(PodOperationsImpl.java:83)
        at cn.com.antcloud.nw.integration.common.tunnel.impl.K8sTunnel.internalUploadFile(K8sTunnel.java:209)
```

Any suggestion to find out the reason?

similar issues: #2217 https://github.com/airbytehq/airbyte/pull/4214

**Comment by wwulfric** [comment]
It seems happening in some specific pods. But I don't know why. Bad pods do not have bash inside, while normal pods do.

**Comment by shawkins** [comment]
See #3384

**Comment by wwulfric** [comment]
@shawkins Yes I found it and gave  a similar title. The file is not uploaded successfully in my case.

**Comment by shawkins** [comment]
Okay, so ignoring the logging of the npes, you are asking more about the exception:

```
io.fabric8.kubernetes.client.KubernetesClientException: [text={"metadata":{},"status":"Success"}]
```

That appears to be an issue PodUploadWebSocketListener.onMessage - any message larger than a single byte is assumed to be an error.  It would be good it include the relevant code and your kubernetes version.  I'm not familiar enough to say offhand what the expectation here is - @rohanKanojia any idea?

**Comment by rohanKanojia** [comment]
I'm sorry I don't know about it. I think @manusa  can provide better insight since he implemented upload funtionality.

**Comment by wwulfric** [comment]
@shawkins What's the kubectl command of sdk upload? Maybe I can try upload from kubectl

**Comment by shawkins** [comment]
https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#cp

**Comment by wwulfric** [comment]
I write a simple demo jar
```
String kubeConfig = "...";
Config config = Config.fromKubeconfig(kubeConfig);
KubernetesClient client = new DefaultKubernetesClient(config);
Path path = new File(filePath).toPath();
client.pods().inNamespace(...).withName(...).inContainer(...)
                .file(...).upload(path);
```
It runs normal in my mac, but fails in my business pod with exception:

```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Sat Sep 18 10:36:18 CST 2021
Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: PodUploadWebSocketListener failed with no response
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUploadWebSocketListener.checkError(PodUploadWebSocketListener.java:91)
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUploadWebSocketListener.waitUntilComplete(PodUploadWebSocketListener.java:111)
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUpload.uploadFile(PodUpload.java:87)
        at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUpload.upload(PodUpload.java:64)
        at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.lambda$upload$0(PodOperationsImpl.java:389)
        at io.fabric8.kubernetes.client.utils.OptionalDependencyWrapper.wrapRunWithOptionalDependency(OptionalDependencyWrapper.java:39)
        at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.upload(PodOperationsImpl.java:387)
        at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.upload(PodOperationsImpl.java:83)
        at FabricK8sClientTest.main(FabricK8sClientTest.java:47)
```

Notice that the file is uploaded successfully.

**Comment by manusa** [comment]
> It runs normally in my mac, but fails in my business pod with exception:

Can it be a networking error? Did you manage to make it work with kubectl (in you business pod)?

**Comment by wwulfric** [comment]
@manusa I've tried many times. It works from my laptop with kubectl and jar. It fails from my business pod with jar.  It works from one master node with kubectl.

**Comment by prathi-adeptia** [comment]
I have the same problem. I tried the official library from Kubernetes and it worked fine without any problems. The file is created but the content in file is missing. Throughs an exception as well: `io.fabric8.kubernetes.client.KubernetesClientException: [text={"metadata":{},"status":"Success"}]`

**Comment by shawkins** [comment]
The pod upload logic uses it's own websocket listener, rather than reusing the more general exec listener.

The exception is originating from https://github.com/fabric8io/kubernetes-client/blob/7ef49cd34e2de0d68422e17ce24fe15e48b63f17/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/uploadable/PodUploadWebSocketListener.java#L55

The assumption being that the first byte received (2 - the stderr stream key) will signify ready.  If a larger message is ever seen, it's treated as an error - the first byte is not being checked as to the response stream.

The pod upload logic is only requesting stderr output, so the logic assumes that it will not see anything on stream 1 (stdout).  And any message seen on 2 or 3 (connection error stream) is assumed to be an error.

The guess is that a proxy or something is sending what you see in error on the stdout stream - which is not expected.

In another client, the upload logic is derived from the exec listener, so it would then effectively ignore anything sent on the stdout stream.

Could you debug and verify this is the case - that the offending response has a first byte of 1 - and if possible determine what is sending this when stdout has not been requested?

**Comment by shawkins** [comment]
> Could you debug and verify this is the case - that the offending response has a first byte of 1 - and if possible determine what is sending this when stdout has not been requested?

Based upon #3793 it seems possible that the first byte is a 3 - and that the interpretation of it as an error is not correct.

I think the best course forward regardless is that after #3793 is addressed that the pod upload listener is modified / removed to be based upon the general exec listener instead.

**Comment by shawkins** [comment]
Beyond the npe, which had already been addressed, I'm seeing three issues listed above:
- a reported failure based upon a 0 exit code status sent to channel 3 - addressed by #3799
- a "no response" error without any details - #3799 will provide the stacktrace, but I don't know the cause.
- the file is created, but there is no content

I was not able to reproduce any of these directly.  It seems like there could be an api server version that responds to a close message by sending the exit code prior to closing, but I didn't see that behavior locally.
